src/lib/device/journal_vdev.cpp

```diff
@@ -559,54 +574,108 @@ void JournalVirtualDev::Descriptor::truncate(off_t truncate_offset) {
     }
 
     // Find the chunk which has the truncation offset. This will be the new
-    // head chunk in the list. We first update the is_head is true of this chunk.
+    // head chunk in the list. We first update the is_head to true for that new head chunk.
     // So if a crash happens after this, we could have two chunks which has is_head
     // true in the list and during recovery we select head with the highest creation
-    // timestamp and reuse or cleanup the other.
-    auto [new_head_chunk, _, offset_in_chunk] = offset_to_chunk(truncate_offset);
+    // timestamp and cleanup the other. We check if the truncation offset happens to be
+    // between the end_of_chunk mark and actual chunk end. In that case there is not data
+    // in that chunk and we release this chunk and select the next chunk.
+    bool update_truncate_offset = false;
+    auto [new_head_chunk, index, offset_in_chunk] = offset_to_chunk(truncate_offset);
+    auto end_new_head_chunk = m_vdev.get_end_of_chunk(new_head_chunk);
+    if (offset_in_chunk >= static_cast< int64_t >(end_new_head_chunk)) {
+        // If the truncation offset is same as end_of_chunk, we dont have any more
+        // data in this chunk to read. Select the next chunk.
+        if (++index <= (m_journal_chunks.size() - 1)) {
+            // Go to the next chunk to make it the new head chunk.
+            new_head_chunk = m_journal_chunks[index];
+            update_truncate_offset = true;
+            size_to_truncate += (new_head_chunk->size() - end_new_head_chunk);
+        }
+    }
+
     auto* private_data = r_cast< JournalChunkPrivate* >(const_cast< uint8_t* >(new_head_chunk->user_private()));
     private_data->is_head = true;
     private_data->logdev_id = m_logdev_id;
     m_vdev.update_chunk_private(new_head_chunk, private_data);
 
     // Find all chunks which needs to be removed from the start of m_journal_chunks.
     // We stop till the truncation offset. Start from the old data_start_offset.
-    // Align the data_start_offset to the chunk_size as we deleting chunks and
-    // all chunks are same size in a journal vdev.
-    uint32_t start = sisl::round_down(ds_off, m_vdev.info().chunk_size);
+    // All the chunks which are covered inside the truncate offset are released.
+    // cover_offset is a logical offset.
+    index = 0;
+    off_t tail_off =
+        static_cast< off_t >(data_start_offset() + m_write_sz_in_total.load(std::memory_order_relaxed)) + m_reserved_sz;
+    auto chunk_size = m_vdev.info().chunk_size;
+    LOGINFOMOD(journalvdev, "Truncate begin truncate {} desc {}", to_hex(truncate_offset), to_string());
```

Comment:
I think we need to log it as periodic log and make it info level, in production. We can do every_10th of logging if afraid to much logs.

Reply:
Yeah periodic log better. But is there a way to do periodic log to log every time when we run UT ? Was hard to debug issues because it wont log every during UT runs.

Reply:
we can just move this log to periodic log and log every time.

Reply:
done.